File: packages/guards/src/credit-info.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type CreditEntity =
  | 'bank-account'
  | 'card-number'
  | 'credit-score';

interface CreditInfoOptions {
  entities: CreditEntity[];
  action: 'block' | 'warn' | 'mask';
}

interface CreditMatch {
  type: CreditEntity;
  value: string;
  start: number;
  end: number;
}

const BANK_ACCOUNT_RE =
  /\d{3,6}-\d{2,6}-\d{4,6}/g;

const CARD_NUMBER_RE =
  /\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b/g;

const CREDIT_SCORE_RE =
  /(?:신용등급|신용점수|크레딧스코어).{0,30}/g;

const MASK_LABELS: Record<CreditEntity, string> = {
  'bank-account': '[계좌번호]',
  'card-number': '[카드번호]',
  'credit-score': '[신용정보]',
};

function detect(
  text: string,
  entities: CreditEntity[],
): CreditMatch[] {
  const matches: CreditMatch[] = [];

  for (const entity of entities) {
    let re: RegExp;
    switch (entity) {
      case 'bank-account':
        re = new RegExp(BANK_ACCOUNT_RE.source, 'g');
        break;
      case 'card-number':
        re = new RegExp(CARD_NUMBER_RE.source, 'g');
        break;
      case 'credit-score':
        re = new RegExp(CREDIT_SCORE_RE.source, 'g');
        break;
    }

    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      matches.push({
        type: entity,
        value: m[0],
        start: m.index,
        end: m.index + m[0].length,
      });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function maskText(
  text: string,
  matches: CreditMatch[],
): string {
  let result = text;
  for (const m of matches) {
    result =
      result.slice(0, m.start) +
      MASK_LABELS[m.type] +
      result.slice(m.end);
  }
  return result;
}

export function creditInfo(
  options: CreditInfoOptions,
): Guard {
  return {
    name: 'credit-info',
    version: '0.1.0',
    description: 'Korean financial info detection',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, options.entities);
      const triggered = matches.length > 0;

      if (!triggered) {
        return {
          guardName: 'credit-info',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'credit-info',
          passed: true,
          action: 'override',
          overrideText: maskText(text, matches),
          latencyMs: Math.round(performance.now() - start),
          details: {
            detected: matches.map(({ type, value }) => ({
              type,
              value,
            })),
          },
        };
      }

      return {
        guardName: 'credit-info',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: {
          detected: matches.map(({ type, value }) => ({
            type,
            value,
          })),
        },
      };
    },
  };
}

File: python/open_guardrail/guards/credit_info.py
"""Credit card number detection and masking guard."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "visa": re.compile(r"4\d{15}"),
    "mastercard": re.compile(r"5[1-5]\d{14}"),
    "amex": re.compile(r"3[47]\d{13}"),
}

_MASK_LABELS: dict[str, str] = {
    "visa": "[CREDIT-CARD]",
    "mastercard": "[CREDIT-CARD]",
    "amex": "[CREDIT-CARD]",
}


def _detect(
    text: str, entities: list[str]
) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append(
                {
                    "type": entity,
                    "value": m.group(),
                    "start": m.start(),
                    "end": m.end(),
                }
            )
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask_text(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        label = _MASK_LABELS.get(m["type"], "[PII]")
        result = (
            result[: m["start"]] + label + result[m["end"] :]
        )
    return result


class _CreditInfo:
    def __init__(
        self,
        *,
        action: str = "block",
        entities: Optional[List[str]] = None,
    ) -> None:
        self.name = "credit-info"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000

        if not matches:
            return GuardResult(
                guard_name="credit-info",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        detected = [
            {"type": m["type"], "value": m["value"]}
            for m in matches
        ]

        if self.action == "mask":
            return GuardResult(
                guard_name="credit-info",
                passed=True,
                action="override",
                override_text=_mask_text(text, matches),
                latency_ms=round(elapsed, 2),
                details={"detected": detected},
            )

        return GuardResult(
            guard_name="credit-info",
            passed=False,
            action=self.action,
            message="Credit card number detected",
            latency_ms=round(elapsed, 2),
            details={
                "detected": detected,
                "reason": (
                    "Text contains credit card numbers"
                ),
            },
        )


def credit_info(
    *,
    action: str = "block",
    entities: Optional[List[str]] = None,
) -> _CreditInfo:
    return _CreditInfo(action=action, entities=entities)
